src/pybind/mgr/rbd_support/module.py

```diff
@@ -483,13 +501,735 @@ def get_perf_counters(self, pool_spec, sort_by):
+                ioctx.remove_omap_keys(write_op, omap_keys)
+                ioctx.operate_write_op(write_op, RBD_TASK_OID)
+        except rados.ObjectNotFound:
+            pass
+
+        if remove_in_memory:
+            try:
+                del self.tasks_by_id[task.task_id]
+                del self.tasks_by_sequence[task.sequence]
+            except KeyError:
+                pass
+
+    def execute_task(self, sequence):
+        task = self.tasks_by_sequence[sequence]
+        self.log.info("execute_task: task={}".format(str(task)))
+
+        try:
+            with self.open_ioctx((task.refs[TASK_REF_POOL_NAME],
+                                  task.refs[TASK_REF_POOL_NAMESPACE])) as ioctx:
+                action = task.refs[TASK_REF_ACTION]
+                execute_fn = {TASK_REF_ACTION_FLATTEN: self.execute_flatten,
+                              TASK_REF_ACTION_REMOVE: self.execute_remove,
+                              TASK_REF_ACTION_TRASH_REMOVE: self.execute_trash_remove,
+                              TASK_REF_ACTION_MIGRATION_EXECUTE: self.execute_migration_execute,
+                              TASK_REF_ACTION_MIGRATION_COMMIT: self.execute_migration_commit,
+                              TASK_REF_ACTION_MIGRATION_ABORT: self.execute_migration_abort
+                              }.get(action)
+                if not execute_fn:
+                    self.log.error("Invalid task action: {}".format(action))
+                else:
+                    task.in_progress = True
+                    self.in_progress_task = task
+                    self.update_progress(task, 0)
+
+                    self.lock.release()
+                    try:
+                        execute_fn(ioctx, task)
+                    except rbd.OperationCanceled:
+                        self.log.info("Operation canceled: task={}".format(
+                            str(task)))
+                    finally:
+                        self.lock.acquire()
+
+                    task.in_progress = False
+                    self.in_progress_task = None
+
+                    self.complete_progress(task)
+                    self.remove_task(ioctx, task)
+
+        except (rados.Error, rbd.Error) as e:
+            self.log.error("execute_task: {}".format(e))
+            self.update_progress(task, 0)
+        finally:
+            task.in_progress = False
+            task.retry_time = datetime.now() + TASK_RETRY_INTERVAL
+
+    def progress_callback(self, task, current, total):
+        progress = float(current) / float(total)
+        self.log.debug("progress_callback: task={}, progress={}".format(
+            str(task), progress))
+
+        # avoid deadlocking when a new command comes in during a progress callback
+        if not self.lock.acquire(False):
+            return 0
+
+        try:
+            if not self.in_progress_task or self.in_progress_task.canceled:
+                return -errno.ECANCELED
```

Comment:
python2 does not knows about `errno.ECANCELED`. Due to the exception was catched, the error did not propagate and `progeress_callback` returned 0, so cancel did not work. To proceed with testing I changed it to -125 locally.

Reply:
Since I already had to add `ECANCELED` to `rbd.pyx`, I think I'll just re-export it so that I don't need to hard-code a constant.